src/dtype.rs

```diff
@@ -94,12 +106,202 @@ impl PyArrayDescr {
+    }
+
+    /// Returns true if the dtype is a sub-array at the top level.
+    ///
+    /// Equivalent to [`np.dtype.hasobject`](https://numpy.org/doc/stable/reference/generated/numpy.dtype.hasobject.html)
+    pub fn has_object(&self) -> bool {
+        self.flags() & NPY_ITEM_HASOBJECT != 0
+    }
+
+    /// Returns true if the dtype is a struct which maintains field alignment.
+    ///
+    /// This flag is sticky, so when combining multiple structs together, it is preserved
+    /// and produces new dtypes which are also aligned.
+    ///
+    /// Equivalent to [`np.dtype.isalignedstruct`](https://numpy.org/doc/stable/reference/generated/numpy.dtype.isalignedstruct.html)
+    pub fn is_aligned_struct(&self) -> bool {
+        self.flags() & NPY_ALIGNED_STRUCT != 0
+    }
+
+    /// Returns true if the data type is a sub-array.
+    pub fn has_subarray(&self) -> bool {
+        // equivalent to PyDataType_HASSUBARRAY(self)
+        unsafe { !(*self.as_dtype_ptr()).subarray.is_null() }
+    }
+
+    /// Returns true if the data type is a structured type.
+    pub fn has_fields(&self) -> bool {
+        // equivalent to PyDataType_HASFIELDS(self)
+        unsafe { !(*self.as_dtype_ptr()).names.is_null() }
+    }
+
+    /// Returns true if the data type is unsized
+    pub fn is_unsized(&self) -> bool {
+        // equivalent to PyDataType_ISUNSIZED(self)
+        self.itemsize() == 0 && !self.has_fields()
+    }
+
+    /// Returns true if data type byteorder is native, or `None` if not applicable.
+    pub fn is_native_byteorder(&self) -> Option<bool> {
+        // based on PyArray_ISNBO(self->byteorder)
+        match self.byteorder() {
+            b'=' => Some(true),
+            b'|' => None,
+            byteorder if byteorder == NPY_BYTEORDER_CHAR::NPY_NATBYTE as u8 => Some(true),
+            _ => Some(false),
+        }
+    }
+
+    /// Returns an ordered list of field names, or `None` if there are no fields.
+    ///
+    /// The names are ordered according to increasing byte offset.
+    ///
+    /// Equivalent to [`np.dtype.names`](https://numpy.org/doc/stable/reference/generated/numpy.dtype.names.html).
+    pub fn names(&self) -> Option<Vec<&str>> {
+        if !self.has_fields() {
+            return None;
+        }
+        let names = unsafe { PyTuple::from_borrowed_ptr(self.py(), (*self.as_dtype_ptr()).names) };
+        <_>::extract(names).ok()
+    }
+
+    /// Returns a dictionary of fields, or `None` if not a structured type.
+    ///
+    /// The dictionary is indexed by keys that are the names of the fields. Each entry in
+    /// the dictionary is a tuple fully describing the field: `(dtype, offset)`.
+    ///
+    /// Note: titles (the optional 3rd tuple element) are ignored.
+    ///
+    /// Equivalent to [`np.dtype.fields`](https://numpy.org/doc/stable/reference/generated/numpy.dtype.fields.html).
+    pub fn fields(&self) -> Option<BTreeMap<&str, (&PyArrayDescr, usize)>> {
```

Comment:
Questions here:

- Should it be an `Option<BTreeMap>`?
- Alternatively, it could be an `Option<impl Iterator<Item = (&str, (&PyArrayDescr, usize))> + '_>`, any better?
- NumPy sort-of-guarantees that those fields will be correct tuples of correct size, maybe it's ok to just `.unwrap()`?

Reply:
I think the iterator would either require unwrapping or having `Result` as the item type. It feels overcomplicated to me. Personally, I am admittedly not to happy about the additional API surface in the first, especially a lot of this duplicates macros which would change theoretically upstream without us noticing.

Reply:
If we do the unpacking, then `Option<impl Iterator>` seems preferable as it makes fewer decisions for calling code (which for example might not care about ordering).

Returning `impl Trait` in a public API might be considered bad style though is it very difficult to store such a type anywhere (as it cannot be named), but I think this not the most pressing concern here as the iterator has a limited lifetime derived from `&self`.

But what I actually wonder is whether we shouldn't just return `&PyDict` instead of doing the unpacking ourselves. Do I understand it correctly that this is to simplify test code where working directly with `PyDict` and `PyTuple` etc. is rather unergonomic?

Reply:
>  additional API surface in the first

Do you mean the behaviour and contents of `->fields`? 

That exact getter wasn't changed in the last 16 years at least since that's where the explicit getter was introduced and the immutable dictproxy was added ([link](https://github.com/numpy/numpy/commit/95fc4486e429c82a872f3debcf09b0273fe2523e)), and it existed in the same form, but less formally, and before that. Given the amount of external stuff relying on numpy, it's pretty much set in stone by now.

If you mean the 'additional API surface' in general - almost all of these getters simply expose various fields of `PyArray_Descr` without any alterations, providing the same API (or close to that) as the numpy itself. I'm not sure I understand.

Reply:
> That exact getter wasn't changed in the last 16 years at least since that's where the explicit getter was introduced and the immutable dictproxy was added (link), and it existed in the same form, but less formally, and before that. Given the amount of external stuff relying on numpy, it's pretty much set in stone by now.

Tried to capture this by "theoretically" as C code using the macros which is not rebuilt would have the same issue with such changes.

> If you mean the 'additional API surface' in general - almost all of these getters simply expose various fields of PyArray_Descr without any alterations, providing the same API (or close to that) as the numpy itself. I'm not sure I understand.

They still raise design questions like the `impl Iterator` or `BTreeMap` choice. I also fear that this API will see little usage outside of special circumstances where the FFI API might provide a reasonable alternative, i.e. probably not when processing arrays in Python extensions.

Reply:
@adamreichold @davidhewitt Would everyone be happy with the suggestion in https://github.com/PyO3/rust-numpy/pull/261#issuecomment-1016958418? If yes, I'll replace `fields` with `get_field` which will keep the API leaner and proceed on with writing tests for all of this. This will be the only `get_`-prefixed method then, which should make it easier to notice it does something less trivial than the other ones.

Reply:
Yes, I like the solution of exposing `get_field` instead of `fields`.